### PR TITLE
HELL YEAH DATABASES

### PR DESCRIPTION
**Pull request by TurboRigby**

### CODE REVIEW

File: options/app_servers.go [hunk 1]
```diff
@@ -13,6 +13,33 @@ type IncludeServers struct {
 	Databases   bool `param:"databases"`   // List of databases on the server
 }
 
+type ListDatabasesOptions struct {
```
Comment (m41denx) [inline options/app_servers.go:40]: type `ListDatabaseOptions` doesn't implement requestOptions and therefore ignores all fields except `Include`

Correct implementation ref:
```go
type ListNodesOptions struct {
	requestOptions
	Include IncludeNodes
}

func (o *ListNodesOptions) getOptions() *requestOptions {
	return &requestOptions{
		Include: o.Include,
	}
}
```

File: options/options.go [hunk 2]
```diff
@@ -14,6 +14,11 @@ type requestOptions struct {
 	SortBy     string
 }
 
+// getOptions implements options.
+func (r *requestOptions) getOptions() *requestOptions {
```
Comment (m41denx) [inline options/options.go]: type `requestOptions` must NOT implement `getOptions()` as it doesn't implement `options` interface on its own

File: options/app_servers.go [hunk 3]
```diff
@@ -13,6 +13,33 @@ type IncludeServers struct {
...
+
+func (o *ListDatabasesOptions) GetOptions() *requestOptions {
+	return &requestOptions{
+		Include: o.Include,
+	}
+}
+
+type GetDatabaseOptions ListDatabasesOptions
+
+func (o *GetDatabaseOptions) GetOptions() *requestOptions {
```
Comment (m41denx) [inline options/app_servers.go:74]: type `GetDatabaseOptions` doesn't implement `requestOptions` and therefore ignores all fields except `Include`

Correct implementation ref:
```go
type ListNodesOptions struct {
	requestOptions
	Include IncludeNodes
}

func (o *ListNodesOptions) getOptions() *requestOptions {
	return &requestOptions{
		Include: o.Include,
	}
}
```

File: app_users.go [hunk 4]
```diff
@@ -24,10 +25,15 @@ type User struct {
 	Servers    []*AppServer `json:"-"`
 }
 
+// FullName returns the full name of the user by concatenating the first name
+// and last name with a space in between.
```
Comment (m41denx) [inline app_users.go:29]: Please drop excessive comments. They do not make your PR look better

File: alligator.go [hunk 5]
```diff
@@ -78,6 +93,9 @@ func (c *Client) newRequest(method, path string, body io.Reader) *http.Request {
 	return req
 }
 
+// validate validates the response from the API request. If the status code is 200, 201, or 202, the response body is read and returned as a byte slice.
```
Comment (m41denx) [inline alligator.go:96]: Please remove your AI-generated comments, they do NOT make godoc good

File: app_databases.go [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/m41denx/alligator/options"
+)
+
+type DatabaseManager struct {
```
Comment (m41denx) [inline app_databases.go:12]: `DatabaseManager` doesn't follow package structure. All methods should be on `App` type itself

File: app_databases.go [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+
+type DatabaseRotatePasswordResponse struct {
+	Password string `json:"password"`
+}
+
+// HELL YEAH 100 DATABSES IN ONE SERVER
+func (dm *DatabaseManager) ListDatabases(serverID int, opts ...options.ListDatabasesOptions) ([]*Database, error) {
+	var o string
+	if opts != nil && len(opts) > 0 {
+		o = options.ParseRequestOptions(opts[0].GetOptions())
```
Comment (m41denx) [inline app_databases.go]:
```go
o = options.ParseRequestOptions(&opts[0])
```

File: app_servers.go [hunk 8]
```diff
@@ -54,15 +60,22 @@ func (s *AppServer) BuildDescriptor() *ServerBuildDescriptor {
...
 }
 
+// DetailsDescriptor creates a ServerDetailsDescriptor for the AppServer.
+// It extracts relevant server details, such as the external ID, name, user ID,
+// and description from the server object.
+// This descriptor is used to initialize or modify server details.
 func (s *AppServer) DetailsDescriptor() *ServerDetailsDescriptor {
 	return &ServerDetailsDescriptor{
 		ExternalID:  s.ExternalID,
-		Name:        s.Name,
```
Comment (m41denx) [inline app_servers.go:60]: Why the hell was it removed??? It is a part of pterodactyl response as per https://dashflo.net/docs/api/pterodactyl/v1/#req_d2df97392474485dbff4a18dcf636495

File: app_servers.go [hunk 9]
```diff
@@ -72,7 +85,207 @@ func (s *AppServer) StartupDescriptor() *ServerStartupDescriptor {
...
+//
+// Parameters:
+//   - serverID: The ID of the server associated with the database.
+//   - databaseID: The ID of the database to retrieve.
+//   - opts: Variadic GetDatabaseOptions for customizing the request.
+//
+// Returns:
+//   - A pointer to the Database object representing the requested database.
+//   - An error if the request fails or the response cannot be parsed.
+func (a *Application) GetDatabase(serverID, databaseID int, opts ...options.GetDatabaseOptions) (*Database, error) {
```
Comment (m41denx) [inline app_servers.go]: I wouldn't even question how this method got here, when it should be part of `app_databases.go`

File: app_servers.go [hunk 10]
```diff
@@ -106,9 +319,17 @@ type ResponseServer struct {
 				Attributes *EggVariable `json:"attributes"`
 			} `json:"data"`
 		} `json:"variables"`
+		Databases struct { // Добавляем структуру для баз данных
```
Comment (m41denx) [inline app_servers.go:285]: That' why my code coverage AI detector goes beeep 67%

**Review by m41denx — CHANGES_REQUESTED** [review]
Looks like shit tbh, rejected for now

File: app_databases.go (hunk 7 above)
Comment (TurboRigby) [reply app_databases.go]: fixed (maybe)

File: app_servers.go (hunk 8 above)
Comment (TurboRigby) [reply app_servers.go:60]: fixed

File: app_servers.go (hunk 9 above)
Comment (TurboRigby) [reply app_servers.go]: fixed (maybe)

File: options/options.go (hunk 2 above)
Comment (TurboRigby) [reply options/options.go]: fixed

**Comment by m41denx** [comment]
Just no

**Pull request closed without merging**